fixes #921: Updated apoc.cypher.doIt with config map and retry feature.

alexiudice opened the pull request:
Added optional config map to `apoc.cypher.doIt`. Currently only takes in a `retries` option that specifies the number of times the cypher query should be retried.

jexp commented on src/main/java/apoc/cypher/Cypher.java:
```diff
@@ -436,4 +445,39 @@ private static Collection asCollection(Object value) {
     public Stream<MapResult> doWhenCase(@Name("conditionals") List<Object> conditionals, @Name(value="elseQuery", defaultValue = "") String elseQuery, @Name(value="params", defaultValue = "") Map<String, Object> params) {
         return whenCase(conditionals, elseQuery, params);
     }
+
```
shouldn't this go somewhere else?
I think I have similar code already for periodic.iterate?

Also I'm not sure that 

1. you get an error from the pooled execution as it usually just kills the thread in the execution, most likely at Future.get and then you get an InvocationTargetException where you have unpack the nested exception
1. it is a transient error? Not many errors make sense to retry, mostly deadlocks only or if some pre-codition is not met yet.

alexiudice replied:
Yes the code was taken from periodic.iterate and was repurposed for `apoc.cypher` calls. The periodic.iterate method I believe you are referring to is`iterateAndExecuteBatchedInSeparateThread`. It has parameters, such as batching and parallel processing which do not make sense in this context. This PR looks to allow a configurable retry. 

I can move this `runWithRetry` method to a Utils class if you think it would be better there.

1. Also I didn’t think the exception itself mattered so I didn't think it needed to be unpacked. It just matters that there is an exception and it is caught so a new transaction can be retried. 

2. Though there may be multiple uses, this PR has one specific use I was thinking of. This is useful for generating unique values in a trigger and then retrying if the value is already in use. 
For example, suppose we have a trigger set up called “addUniqueValueToCreatedNodes” that generates a `uuid` that must be unique (has a uniqueness CONSTRAINT on it) and adds it to created nodes. If a node is added to the graph the trigger will start and generate the `uuid` and try to add it.  If the generated `uuid` is already in use, the trigger will fail and the node will not be created. The ability to run the generation of `uuid` and being able to SET it on the node in `apoc.cypher.doIt` will allow for multiple retries in the trigger itself, so the node will be created. 

Here is the example in code:

`CALL apoc.trigger.add("addUniqueValueToCreatedNodes", "UNWIND {createdNodes} AS n CALL apoc.cypher.doIt(\"WITH {node} AS node SET node.uuid = 'uuid:' + apoc.create.uuid()\", {node: n}, {retries: 5}) YIELD value RETURN value", {phase: "after”})`

This would fail after 5 retries, making it much more likely the node would be created compared to the code below where there is no retry and a conflicting `uuid` would fail to create the node.

`CALL apoc.trigger.add("addUniqueValueToCreatedNodes", "UNWIND {createdNodes} AS n WITH n SET n.uuid = 'uuid:' + apoc.create.uuid()", {phase: "before”})`

jexp replied:
Ok, makes sense.

We meant to add an uuid adding feature to apoc anyway though :)

Btw. it's not recommended to uses prefixes like `uuid:` as that affects the ability for string compression as it leaves the character set of hex-code+separator.

alexiudice commented on src/main/java/apoc/cypher/Cypher.java:
```diff
@@ -436,4 +445,39 @@ private static Collection asCollection(Object value) {
+                // Query has not finished yet, wait.
+                while ( !resultFuture.isDone() )
+                {
+                    LockSupport.parkNanos( 1000 );
```
Would a `Util.sleep` suffice here rather than `LockSupport.parkNanos`?

alexiudice commented:
I changed how the transaction was being executed, now the `runWithRetry` method is able to get the type of Neo4j error that occurs. 

For exceptions of type `QueryExecutionException` (the standard type of exception thrown by the graphdb)  I went through the different StatusCodes and created a whitelist for which failed transactions can be retried by default. Like you mentioned many exceptions do not make sense to retry. We can add or remove from the list, it was just a first pass through.
 ```
General.UnknownError,
Network.CommunicationError,
Procedure.ProcedureTimedOut,
Schema.ConstraintValidationFailed,
Schema.SchemaModifiedConcurrently,
Statement.ConstraintVerificationFailed,
Transaction.ConstraintsChanged,
Transaction.DeadlockDetected,
Transaction.InstanceStateChanged,
Transaction.LockAcquisitionTimeout,
Transaction.LockSessionExpired,
Transaction.Outdated,
Transaction.TransactionCommitFailed,
Transaction.TransactionStartFailed,
Transaction.TransactionTimedOut```
(Documentation can be found here: https://neo4j.com/docs/developer-manual/current/reference/status-codes/)

Something like a “SyntaxError” will not retried and the StatusCode of the exception is pushed up to the user.
I also made an optional parameter to pass in StatusCodes, so if you wanted to retry on a SyntaxError (or any other errors not in the whitelist) you could.

Lastly I moved all the methods into the Util class, it seemed like the best place for these methods.

I'm thinking the last thing this PR needs is to add to the documentation and maybe writing some tests. I've tested this method out manually in several contexts and it works. I can create some tests once I know the PR is not going to change too much.

@jexp Anything else that I should change? Did I address your concerns?

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+            } );
+            Thread threadForTransaction = new Thread( resultFuture );
+
+            try
```
can you use the thread pool and not start manual threads?

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+
+    public static Stream<MapResult> runWithRetry( GraphDatabaseService db, Log log, String statement, long retries, List<String> additionalRetryCodes,
+            Map<String,Object> params )
+    {
```
please use the regular apoc code format

alexiudice replied:
Where could I find the apoc code format scheme?

jexp replied:
Regular Java Style guide (e.g. IntelliJ / Eclipse defaults)

alexiudice replied:
I'll update to the regular style guide as my final step. Want to get all the code completed before changing the aesthetics.

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+    private static boolean isRetriableException( String statusCode, List<String> customRetriable )
+    {
+
+        if ( RetriableExceptions.QueryExecutionExceptionStatusCodeList.stream().map( sc -> sc.code().serialize().equals( statusCode ) ).reduce( ( x, acc ) -> acc || x ).get()
```
probably a loop + if with early return is easier here ?
there is also any/none() etc on streams

alexiudice replied:
I'll fix this up.

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+
+    private interface RetriableExceptions extends Status
+    {
+        List<Status> QueryExecutionExceptionStatusCodeList = Arrays.asList(
```
why is this in an interface?
did you check if status has equals perhaps a set with contains is good enough?

jexp replied:
we should utilize Code.classification for transient errrors
https://github.com/neo4j/neo4j/blob/3.5/community/common/src/main/java/org/neo4j/kernel/api/exceptions/Status.java#L80-L91

alexiudice replied:
@jexp It is an interface since the inner enums of Status are `protected`. And I did check the Status interface, it does not have the exceptions grouped in a nice way for what we would want to retry on. We want to have control over which exceptions can cause a retry.

What were you thinking/meaning by utilizing `code.classification` for transient errors?

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+            FutureTask<Result> resultFuture = new FutureTask<>( () -> {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Result result = ((Function<Map<String,Object>,Result>) ( Map<String,Object> p ) -> db.execute( withParamMapping( statement, p.keySet() ),
```
what is wrong with just calling it?

alexiudice replied:
I was thinking that we only want the result once the query has finished. By putting it in a FutureTask I was thinking we would only return the result once the entire query was done. I will remove the FutureTask and just run it.

alexiudice replied:
I was changing it back and I realized why I had it as a FutureTask rather than just calling it. 

I did this because if the Tx fails, it marks every transaction created in the thread for `rollback`. So even if the retry is successful the transaction is always rolled back. 

This is why I put it in a FutureTask object and then ran it in a separate thread each time.

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+                }
+                catch ( Exception e )
+                {
+                    throw e;
```
this doesn't make sense?

jexp commented on src/main/java/apoc/util/Util.java:
```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+                // Query has not finished yet, wait.
+                while ( !resultFuture.isDone() )
+                {
+                    LockSupport.parkNanos( 1000 );
```
1ms is a very long time, I think we should rather stream and apply batches and then handled the failed ones on the side, e.g. by putting them into a queue.

jexp commented:
@alexiudice ping, let me know if/when I should review again.

jexp commented:
@alexiudice ping

alexiudice commented:
Sorry I've been swamped. I haven't forgotten about this!

alexiudice commented on src/main/java/apoc/util/Util.java:
```diff
@@ -700,4 +706,99 @@ public static boolean classExists(String className) {
+                while ( !resultFuture.isDone() )
+                {
+                    LockSupport.parkNanos( 100 );
+                }
```
Does this wait time work? Or do we want to scrap the waiting entirely and try to do something by streaming the result? I'm just not sure how I'd go about doing that.

alexiudice commented:
@jexp ping.

I've updated the PR, though I was unsure of some of the changes you wanted me to make. Some of the comments/questions I left are not showing up on the Files Changed, not sure if you can seem them or you need me to retype them.
